File: Android/app/src/main/java/com/alafia/android/views/hd/TherapyReportPrinter.kt
package com.alafia.android.views.hd

import android.content.Context
import android.print.PrintAttributes
import android.print.PrintManager
import android.webkit.WebView
import android.webkit.WebViewClient
import com.alafia.android.api.ApiClient
import com.alafia.android.util.ErrorUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Print or save a treatment report as PDF.
 *
 * The document is rendered SERVER-SIDE and fetched as HTML, so the patient's
 * copy, the clinician's copy and the web copy are the same document. Three
 * templates would drift, and the one that drifted would be the one a clinician
 * was reading.
 *
 * `WebView.createPrintDocumentAdapter` hands it to Android's own print
 * pipeline, which gives the user "Save as PDF" and any configured printer for
 * free — affordances a bundled PDF library would have taken away.
 */
object TherapyReportPrinter {

    /**
     * Fetch and open the print dialog. Returns an error message, or null on
     * success.
     *
     * [patientId] is set when a CLINICIAN prints; the backend re-checks the
     * sharing grant on that route, so passing an id is a destination and never
     * an authorisation.
     */
    suspend fun print(context: Context, sessionId: Int, patientId: Int? = null): String? {
        val html = try {
            withContext(Dispatchers.IO) {
                val api = ApiClient.getApiService()
                val body = if (patientId != null) {
                    api.patientTherapySessionReport(patientId, sessionId)
                } else {
                    api.therapySessionReport(sessionId)
                }
                body.string()
            }
        } catch (e: Exception) {
            // Named, not swallowed: a print button that silently does nothing
            // is indistinguishable from one that was never wired.
            return ErrorUtil.userMessage(e)
        }

        return withContext(Dispatchers.Main) {
            try {
                // The WebView must be held until the adapter is created, and
                // printing may only start once the content has finished
                // loading — starting earlier prints a blank page.
                val webView = WebView(context)
                webView.webViewClient = object : WebViewClient() {
                    override fun onPageFinished(view: WebView, url: String) {
                        val manager = context.getSystemService(Context.PRINT_SERVICE) as PrintManager
                        manager.print(
                            "ALAFIA Treatment Report",
                            view.createPrintDocumentAdapter("ALAFIA Treatment Report"),
                            PrintAttributes.Builder().build(),
                        )
                    }
                }
                webView.loadDataWithBaseURL(null, html, "text/html", "UTF-8", null)
                null
            } catch (e: Exception) {
                ErrorUtil.userMessage(e)
            }
        }
    }
}

File: IOS/ALAFIA/Views/Hemodialysis/TherapyReportPrinter.swift
import SwiftUI
import UIKit

/// Print or save a treatment report as PDF.
///
/// The document is rendered SERVER-SIDE and fetched as HTML, so the patient's
/// copy, the clinician's copy and the web copy are the same document. Three
/// templates would drift, and the one that drifted would be the one a clinician
/// was reading.
///
/// `UIMarkupTextPrintFormatter` hands the HTML to iOS's own print pipeline,
/// which gives AirPrint, "Save to Files" and "Save as PDF" for free — the
/// affordances a bundled PDF library would have taken away.
enum TherapyReportPrinter {

    /// Fetch and present the print sheet. Returns an error message on failure.
    ///
    /// `patientId` is set when a CLINICIAN is printing; the backend re-checks
    /// the sharing grant on that route, so passing an id is a destination and
    /// never an authorisation.
    @MainActor
    static func present(sessionId: Int, patientId: Int? = nil) async -> String? {
        let path = patientId.map {
            "/clinician-dashboard/patient/\($0)/therapy-sessions/\(sessionId)/report.html"
        } ?? "/chronic/therapy-sessions/\(sessionId)/report.html"

        let html: String
        do {
            html = try await APIClient.shared.getHTML(path)
        } catch {
            // Named, not swallowed. A print button that silently does nothing
            // is indistinguishable from one that is not wired at all.
            return error.localizedDescription
        }

        let info = UIPrintInfo(dictionary: nil)
        info.outputType = .general
        info.jobName = "ALAFIA Treatment Report"

        let controller = UIPrintInteractionController.shared
        controller.printInfo = info
        controller.printFormatter = UIMarkupTextPrintFormatter(markupText: html)

        return await withCheckedContinuation { continuation in
            controller.present(animated: true) { _, _, error in
                continuation.resume(returning: error?.localizedDescription)
            }
        }
    }
}
